src/ck-core/ckarray.C

```diff
@@ -1386,16 +1371,49 @@ void CProxy_ArrayBase::ckBroadcast(CkArrayMessage *msg, int ep, int opts) const
 		else
 			CProxy_CkArray(_aid).recvBroadcast(msg);
 	  } else {
-		DEBB((AA "Forwarding array broadcast to serializer node %d\n" AB,CpvAccess(serializer)));
-		CProxy_CkArray ap(_aid);
-		if (skipsched)
-			ap[CpvAccess(serializer)].sendExpeditedBroadcast(msg);
-		else
-			ap[CpvAccess(serializer)].sendBroadcast(msg);
-	  }
+			DEBB((AA "Forwarding array broadcast to serializer node %d\n" AB,CpvAccess(serializer)));
+			CProxy_CkArray ap(_aid);
+#if CMK_ONESIDED_IMPL
+			if(CMI_ZC_MSGTYPE(env) == CMK_ZC_BCAST_SEND_MSG ||
+				CMI_ZC_MSGTYPE(env) == CMK_ZC_BCAST_RECV_MSG) {
+				
+				// ZC Bcast is implemented on non-zero root nodes by sending a small message to Node 0 (through comm thread)
+				// to increment bcastNo on PE 0 i.e. the serializerPe (implemented as an atomic). After incrementing, an ack message is sent back to
+				// this PE (which is the root node pe) to perform a broadcast
+				MsgPointerWrapper w;
+				w.msg = msg;
+				ap[CpvAccess(serializer)].incrementBcastNoAndSendBack(CkMyPe(), w);
+			} else
+#endif
+			{
+				// Regular Bcast (non ZC) is implemented on non-zero root nodes by forwarding the message to PE 0 and
+				// then having PE 0 perform the broadcast rooted at Node 0. This is done to ensure single delivery (and avoid
+				// multiple delivery or no delivery of the message) when load balancing occurs during a broadcast
+				DEBB((AA "Forwarding array broadcast to serializer node %d\n" AB,CpvAccess(serializer)));
+				if (skipsched)
+					ap[CpvAccess(serializer)].sendExpeditedBroadcast(msg);
+				else
+					ap[CpvAccess(serializer)].sendBroadcast(msg);
+			}
+		}
 	}
 }
 
+void CkArray::incrementBcastNoAndSendBack(int srcPe, MsgPointerWrapper w){
+    int zcBcastNo = broadcaster->incrementBcastNo();
+
+    // Send back to CkArray on that index
+    thisProxy[srcPe].sendZCBroadcast(zcBcastNo, w);
+  }
+
+void CkArray::incrementBcastNo(){
+  int zcBcastNo = broadcaster->incrementBcastNo();
+}
+
+void CkArray::sendZCBroadcast(int zcBcastNo, MsgPointerWrapper w) {
```

Comment:
Why does this take zcBcastNo as an argument if it doesn't use it?

Reply:
Good point, I'll change this.